server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -270,6 +294,52 @@ public ClusterMetadataManifest writeIncrementalMetadata(
         return manifest;
     }
 
+    /**
+     * Uploads provided ClusterState's global Metadata to remote store in parallel.
+     * The call is blocking so the method waits for upload to finish and then return.
+     *
+     * @param clusterState current ClusterState
+     * @return String file name where globalMetadata file is stored.
+     */
+    private String writeGlobalMetadata(ClusterState clusterState)
+        throws IOException {
+
+        AtomicReference<String> result = new AtomicReference<String>();
+        final BlobContainer globalMetadataContainer = globalMetadataContainer(
+            clusterState.getClusterName().value(),
+            clusterState.metadata().clusterUUID()
+        );
+        final String globalMetadataFilename = globalMetadataFileName(clusterState.metadata());
+
+        // latch to wait until upload is not finished
+        CountDownLatch latch = new CountDownLatch(1);
```

Comment:
You can use LatchedActionListener

Reply:
We need to wire the latch with the index metadata in case index metadata and cluster metadata happens in the same cluster state update. We should reuse the existing latch

Reply:
>You can use LatchedActionListener

Used LatchedActionListener

> We need to wire the latch with the index metadata in case index metadata and cluster metadata happens in the same cluster state update. We should reuse the existing latch

I believe wiring same latch in index metadata and global metadata will help when we upload both the things in parallel. In that case we can wait for one latch only, but I think in current implementation both the things are uploaded sequentially only. Do you see any benefit of wiring same latch when both are sequential?

Reply:
Do you think index metadata and cluster metadata can be uploaded concurrently(if such a flow exists)?

Reply:
Yes, This can be optimization for the cases where index metadata and global metadata is updated in single update.

On a second thought, can we change any index metadata and global metadata in single cluster update? I think there is not any API which performs both the update in single cluster state update, Correct me if that's not the case.

Reply:
can some internal flow change both at once? maybe something in Index template or data stream?

Reply:
Also what happens when we have a partial failure i.e. global metadata succeeds but index metadata fails. Would it be possible to expose dirty cluster state?

Reply:
@shwetathareja I have added TODO and created sim for it. [#10645]

@Bukhtawar we write the manifest file to remote store, once both global metadata and index metadata files are successfully uploaded to remote store. In partial case, we wont upload new manifest file. We wont expose dirty cluster state.

Reply:
https://github.com/opensearch-project/OpenSearch/issues/10686